### ISSUE TITLE
(Typescript) Dependency is not reported if the only import is a type. 

### ISSUE THREAD

**Issue by arnauorriols** [sverweij/dependency-cruiser#28]
When only importing a type from a module, dependency-cruiser does not report the dependency. 

### Steps to Reproduce (for bugs)
1. Create module A with interface A
2. Create module B
3. Import interface A from module B
4. Generate dependencies report with dependency-cruiser

### Your Environment
* Version used: 2.10.1
* Node version: v9.3.0
* Operating System and version: Archlinux

**Comment by sverweij** [comment]
@arnauorriols thank you for reporting this. Could you attach the sample
modules you used to reproduce this?

**Comment by arnauorriols** [comment]
Gladly!

**A.ts**
```
import { B } from './B';

export interface A {
	foo: string;
}

const b = new B();
```

**B.ts**
```
import { A } from './A';

export class B {};

const a: A = {foo: "foo"};
```
Expected: bidirectional dependency.

Actual result:
![dependencygraph](https://user-images.githubusercontent.com/4871949/34850975-495a13ae-f728-11e7-8135-a99962960e72.jpeg)

**Comment by sverweij** [comment]
Thanks - I'll take a look at it over the weekend.

**Comment by sverweij** [comment]
@arnauorriols took a bit longer than I anticipated to look at it. 

**TL;DR**
- Root cause: the dependency `B.ts -> A.ts` only exists at compile time, but not after; and dependency-cruiser is using 'after' to derive dependencies.
- Solving this issue is possible, but not easy (as far as I can see).

**What's happening:**
Before inferring dependencies dependency-cruiser first uses the typescript compiler to translate it to (ESNext) javascript. In this step `B.ts` becomes this:

```javascript
export class B {};
const a = { foo: "foo" };
```
So, as far as the typescript compiler is concerned A isn't used and `import { A } from './A';` can be omitted from the output.

**Solution direction**
In order to catch situations like this dependency-cruiser's architecture would need to be changed a bit (at least for typescript); it'd have to use a typescript syntax tree instead of the esnext one. Last time I looked into this (~dec 2016) using the typescript syntax tree was gnarly to work with - but that might have changed in the mean time.

Not a thing I can pick up in the short term (PR welcome, though) other than documenting the behaviour.

**Comment by sverweij** [comment]
@arnauorriols after some research over the weekend:
- your issue is more common than I initially thought
- running through the typescript AST, while still gnarly, seems simpler than it did a year ago 

So I'm going to pick it up. Now. Wish me luck 🤞 

Thanks again for raising this issue!!

**Comment by jbakse** [comment]
I've been trying depcruise out and finding it handy. I think I ran into the same problem as @arnauorriols and this fix _did_ fix it. I wasn't sure if my issue this was the same thing at first.

I had seen a problem where some dependencies were not showing up if an upstream file already imported it. If (A imports B and C) and (B imports C), I wasn't getting an arrow from B to C. If I took the C import away from A, then the B->C import showed up in the graph.

With the beta, things are working like I expected.

Thought I post this over-simplified example as a possible test case.

**a.ts**
```
import { B } from './b';
import { C } from './c';

export class A {
}
```

**b.ts**
```
import { C } from './c';

export class B {
}
```

**c.ts**
```
export class C {
}
```

old
<img width="120" alt="screen shot 2018-01-27 at 2 17 33 pm" src="https://user-images.githubusercontent.com/959425/35475504-a635403e-036d-11e8-9f46-8a36e9f20199.png">

beta
<img width="366" alt="screen shot 2018-01-27 at 2 10 05 pm" src="https://user-images.githubusercontent.com/959425/35475507-a9a6bc84-036d-11e8-817a-4a780f0ceba0.png">


Thanks for your work on this!

**Comment by sverweij** [comment]
@jbakse thanks for your feedback. Glad it works and is of use to you! 

I happened to need a test fixture for this use case, so I copied your example verbatim into the fixture folder (3f35709) and added a snapshot.

As you already surmised, your issue has the same root cause as this one. The typescript compiler weeds out any import that in the javascript output wouldn't be used:

a.js:
```javascript
// import { B } from './b' omitted
// import { C } from './c' omitted
export class A {
}
```

b.js
```javascript
// import { C } from './c' omitted
export class B {
}
```

c.js
```javascript
export class C {
}
```

... which would lead to the 'old' graph when dependency-cruiser would've kept looking at any transpiled output.

(Note that when `a.ts` would start to use class B and C, the dependency would've shown up even in the 'old' situation).

Thanks again for your elaborate feedback!

**Comment by sverweij** [comment]
@arnauorriols / @jbakse / @kivilahtio : dependency-cruiser@latest (currently `2.12.0`) with the fix for this issue is on npmjs since about half an hour. Enjoy!

**Comment by arnauorriols** [comment]
Apologies for not saying anything during the beta stage. I confirm it works with the latest stable. Many thanks!

This issue can be closed from my side.

**Comment by sverweij** [comment]
@arnauorriols no reason to apologize - the beta stage was even shorter than usual (and in the weekend to boot :-) ) .

Good to hear it works for you!

**Comment by sverweij** [comment]
Hi @benlesh, thanks for reporting this! Sorry about the breaking build.

One way to resolve this could be adding a command line switch to
dependency-cruiser to opt in to the new behavior. I can probably add it
tonight (GMT+1).

**Comment by sverweij** [comment]
I've released version 2.13.0 to npm yesterday evening with (a.o.) the contents of the above PR ([release notes](https://github.com/sverweij/dependency-cruiser/releases/tag/v2.13.0)).
